### PR TITLE
More accurate word detection and text selection

### PR DESCRIPTION
**Pull request by poire-z**
Preliminary code for discussion.
Decided to have a go at the em-dash selection problem, and found out a single place where stuff could be solved.
crengine was using only the space character to delimit words (finding start or end of word when onHold or next/prev word when onHoldPan), via this function IsUnicodeSpace(ch), which is used numerous times in the functions folowing the one i modified.
I thought I would have to add a new function IsWordSeparator(ch), and decide which IsUnicodeSpace() calls to replace with this new one, but I first tried to just replace them all by just adding my changes to IsUnicodeSpace()... and the result is quite smooth, with no strange things seen over a few days (tested only on the emulator thus).

With this, word selection stops at space and punctuations, and all signs (hyphen, em-dash, quotations marks, parenthesis...). This means that with the word "open-source", one can hold on `open` to lookup `open`, or hold and pan to `source` to lookup `open-source`, so it allows to lookup what the user decides.
It would also no more need all the stuff I did in `ReaderDictionary:cleanSelection()` to remove dashes, quotations marks, and french grammatical stuff, as they are now no more included in words (but can be included in compound words, on in sentences (but no more at selection edges), by panning to following words).

The side effects are :
- now that words are shorter, the distance to go to the following word is smaller too, and sometimes below the HoldPan minimal distance, so it's hard to select two small words like "open-bar" for highlighting in a small font.
- previously, when highlights included punctuation, when one highlighted a quote `«what a quote !»`, the `«` and `»` were included in the quoted text (seen in bookmarks and in metadata.lua), although they weren't part of the grey highlight boxes (because the code in ldomWordsCollector does not consider them to be words - it did not consider DIGITS to be part of words too, but I corrected that).

For old saved highlights (that included punctuation at edges), this would now cause the previous and next words to be included in the highlight grey box. So, it will need this change in another commit to cre.cpp :
```c
--- a/cre.cpp
+++ b/cre.cpp
@@ -1078,10 +1078,25 @@ static int getWordBoxesFromPositions(lua_State *L) {
                        return 0;
                r.sort();

+               // Old saved highlights may have included punctuation at
+               // edges (they were not displayed in boxes because of
+               // lvtinydom.cpp's ldomWordsCollector catching only letters)
+               // and punctuation was considered part of the word
+               // (a space was the sole word separator), so
+               // r.getStart().isVisibleWordStart() and r.getEnd().isVisibleWordEnd()
+               // were always true.
+               // With our changes to word detection, this is no more true,
+               // and the next tests would gather the previous or following
+               // words, which were not part of the saved highlight !
+               // But we can change the method to indeed get the word just
+               // after or before the punctuation and actually get
+               // exactly the included words without punctuation
                if (!r.getStart().isVisibleWordStart())
-                       r.getStart().prevVisibleWordStart();
+                       // r.getStart().prevVisibleWordStart();
+                       r.getStart().nextVisibleWordStart();
                if (!r.getEnd().isVisibleWordEnd())
-                       r.getEnd().nextVisibleWordEnd();
+                       // r.getEnd().nextVisibleWordEnd();
+                       r.getEnd().prevVisibleWordEnd();
                if (r.isNull())
                        return 0;

```
I had a quick search for issues about highlight, and it looks to me this change would close a few of them (and may be introduce a few new ones ;).

(If we would go with this, you'll have to tell me if I should change the name of the function from IsUnicodeSpace to IsWordSeparator (or something else) everywhere, or if we can stay with that original name for historical/context reasons, and just go with a comment to explain that.)

### CODE REVIEW

**Comment by houqp** [comment]
Sorry that I missed this PR from the mail box. The change looks good to me. I suggest we change the name from IsUnicodeSpace to IsWordSeparator like you mentioned. Functions should be named to help other contributor following the current logic, let's leave the historical/context side of things in the comment ;)

File: crengine/src/lvtinydom.cpp [hunk 1]
```diff
@@ -6553,13 +6553,38 @@ bool ldomXPointerEx::prevVisibleText( bool thisBlockOnly )
...
+    // 0xC0 to 0xFF, except 0xD7 and 0xF7, are latin accentuated letters.
+    // Above 0xFF are other alphabets. Let's consider all above 0xC0 unicode
+    // characters as letters, except the adequately named PUNCTUATION ranges.
+    // There may be exceptions in some alphabets, that we can individually
+    // add here :
+    if (ch == 0xD7 ) return true;  // multiplication sign
+    if (ch == 0xF7 ) return true;  // division sign
+
+    // these includes em-dash & friends, and other quotation marks
+    if (ch>=UNICODE_GENERAL_PUNCTUATION_BEGIN && ch<=UNICODE_GENERAL_PUNCTUATION_END) return true;
```
Comment (Frenzie) [inline crengine/src/lvtinydom.cpp]: Does this also include the various width space characters? See, .e.g., https://www.cs.tut.fi/~jkorpela/chars/spaces.html

Comment (poire-z) [inline crengine/src/lvtinydom.cpp]: Most of them. In include/lvstring.h:
```
#define UNICODE_CJK_PUNCTUATION_BEGIN 0x3000
#define UNICODE_CJK_PUNCTUATION_END 0x303F
#define UNICODE_GENERAL_PUNCTUATION_BEGIN 0x2000
#define UNICODE_GENERAL_PUNCTUATION_END 0x206F
#define UNICODE_CJK_PUNCTUATION_HALF_AND_FULL_WIDTH_BEGIN 0xFF01
#define UNICODE_CJK_PUNCTUATION_HALF_AND_FULL_WIDTH_END 0xFFEE
```
I'll add the remaining ones:
U+1680	OGHAM SPACE MARK
U+180E	MONGOLIAN VOWEL SEPARATOR
U+FEFF	ZERO WIDTH NO-BREAK SPACE
Thanks for the link !
Not sure if I should include the symbols at the bottom of this page.
And there are some other unicode symbols ranges (bullets, icons...) that I guess may not be considered part of a words, but well, we may not see them often in epub. We'll see if that's needed.

Comment (Frenzie) [inline crengine/src/lvtinydom.cpp]: Spaces (at least thin and hair spaces) are actually used, albeit rarely. In proper typesetting an em-dash should be surrounded by thin spaces. Hence why I asked. I don't think we need to worry about bullets and stuff.

**Comment by poire-z** [comment]
Done.
Guess I have to wait for this to be merged before doing a PR to base to bump crengine (I think the commit ref here will be different than mine).
Btw, I had a look at all the diffs since Feb 2014 between original crengine (https://sourceforge.net/p/crengine/crengine/) and our git, crengine/src & include, and I didnt find any interesting fixed or new stuff.

**Comment by houqp** [comment]
Yup, the new changes in coolreader are mostly android changes.

**Comment by virxkane** [comment]
@poire-z Hello again :)
This PR broke sentence selection for TTS in CoolReader android.
If I revert at least one function bool ldomXPointerEx::nextVisibleWordEnd( bool thisBlockOnly ) sentence selection for TTS is restored.
```diff
--- a/crengine/src/lvtinydom.cpp
+++ b/crengine/src/lvtinydom.cpp
@@ -11425,7 +11425,7 @@ bool ldomXPointerEx::nextVisibleWordEnd( bool thisBlockOnly )
         bool nonSpaceFound = false;
         // skip non-spaces
         while ( _data->getOffset()<textLen ) {
-            if ( IsWordSeparator(text[ _data->getOffset() ]) )
+            if ( IsUnicodeSpace(text[ _data->getOffset() ]) )
                 break;
             nonSpaceFound = true;
             _data->addOffset(1);
@@ -11435,13 +11435,13 @@ bool ldomXPointerEx::nextVisibleWordEnd( bool thisBlockOnly )
         if ( nonSpaceFound )
             return true;
         // skip spaces
-        while ( _data->getOffset()<textLen && IsWordSeparator(text[ _data->getOffset() ]) ) {
+        while ( _data->getOffset()<textLen && IsUnicodeSpace(text[ _data->getOffset() ]) ) {
             _data->addOffset(1);
             //moved = true;
         }
         // skip non-spaces
         while ( _data->getOffset()<textLen ) {
-            if ( IsWordSeparator(text[ _data->getOffset() ]) )
+            if ( IsUnicodeSpace(text[ _data->getOffset() ]) )
                 break;
             nonSpaceFound = true;
             _data->addOffset(1);
```
The problem is that sentence completion characters are not captured.
If I revert only this function, what will I break instead?

**Comment by poire-z** [comment]
You're sure it's this PR ? It's very old :) and you must have picked it months ago.
Another more recent PR that touched these methods is https://github.com/koreader/crengine/commit/0dee202b5 #347 .

> If I revert only this function, what will I break instead?

You might break word selection with KOReader. I dunno about CoolReader.
From a quick glance at this patch, I guess the idea was that, may be, with this text:
`It is open-source. You can look.`
Previously, long-pressing on "open" would have `open-source.` selected (with the dot at end).
With this PR I guess we just get `open` - and long-pressing on source, we get `source` (without the dot).
Which I guess is nearer to the meaning of "visible word end" - to not have the dot.

Dunno how that works with CoolReader - or if this function really did have another meaning :/
(like get up to next visible word start, then rewing skipping spaces ?)

> The problem is that sentence completion characters are not captured.

Also note my comment at the top of : 
https://github.com/koreader/crengine/commit/727a0d2f8ca8eb7d08b7e926252155edbbe53ebc
May be it needs some more work on the SentenceEnd/Start functions...

I'm sorry I can't help more, I really hate these next/prevVisibleNightmare functions :) And I really don't want to look at CoolReader and the next/prevVisibleSentences :/

**Comment by poire-z** [comment]
Or may be, to avoid diverging too much:
- duplicate this function to `nextVisibleWordEndAndABitMore()` with your stuff reverted.
- find out who was calling it (I guess, the VisibleSentence methods ?) and have just them call the duplicated function ?

**Comment by virxkane** [comment]
> * duplicate this function to `nextVisibleWordEndAndABitMore()` with your stuff reverted.

I thought about the same.


> * find out who was calling it (I guess, the VisibleSentence methods ?) and have just them call the duplicated function ?

That's where I started, otherwise how did I find the cause of the problem?
```
libcr3engine-3-2-X.so`ldomXPointerEx::nextVisibleWordEnd(this=0x000075be2ee75900, thisBlockOnly=true) at lvtinydom.cpp:11428
libcr3engine-3-2-X.so`ldomXPointerEx::thisSentenceEnd(this=0x000075be2ee75900) at lvtinydom.cpp:11681
libcr3engine-3-2-X.so`LVDocView::onSelectionCommand(this=0x000075be2e19f280, cmd=131, param=0) at lvdocview.cpp:6077
libcr3engine-3-2-X.so`LVDocView::doCommand(this=0x000075be2e19f280, cmd=DCMD_SELECT_FIRST_SENTENCE, param=0) at lvdocview.cpp:5952
libcr3engine-3-2-X.so`::Java_org_coolreader_crengine_DocView_moveSelectionInternal(_env=0x000075be3e07e620, _this=0x000075be2ee75fe4, _sel=0x000075be2ee75fe8, _cmd=131, _param=0) at docview.cpp:1981
```
But now you will say that we do not use any LVDocCmd stuff in the koreader :)

**Comment by poire-z** [comment]
> But now you will say that we do not use any LVDocCmd stuff in the koreader :)

(Yep, fortunatly for me, we don't :) something else I don't have to get into :)

**Pull request merged**